[Mono] Fix warnings to enable W4 on Windows.

lateralusX opened the pull request:
Upstream work done in different downstream repros.

Fix selected W4 warnings on Windows as well as suppressing specific W4 warnings to be fixed. Give us option to do Mono runtime build using W4 on Windows and also turn on warnings as errors to prevent regressions on the warnings we enable.

Fix Warning C4146 in Mono runtime build.

Enable "Treat Warnings As Errors" to prevent regressions on the warnings we decided to fix.

lateralusX committed:
Fix diagnostics_tracing_objects to build under W4 on Windows.

lateralusX committed:
Enable warning level W4 on Windows + disable not yet fixed w4 warnings.

AaronRobinsonMSFT commented:
@lateralusX Is this helping to contribute to https://github.com/dotnet/runtime/issues/66154?

AaronRobinsonMSFT commented:
If so, I would also add the following to help ensure regressions are caught early.

```
add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/WX>) # treat warnings as errors
```

lateralusX commented:
> If so, I would also add the following to help ensure regressions are caught early.
> 
> ```
> add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/WX>) # treat warnings as errors
> ```

Yes, that was my plan to do in this PR, once I worked it through.

lateralusX commented:
> @lateralusX Is this helping to contribute to #66154?

This PR is mainly upstream other work getting W4 building on Mono, so not explicitly target any of those warnings, but yet another stepping stone bringing down warnings altogether on Mono while still bumping to W4.

lateralusX committed:
Fix signed/unsigned in for blocks of modified files.

lateralusX committed:
Fix Warning 4146 on Windows.

lateralusX committed:
Enable warnings as errors on Windows build to prevent regression.

lateralusX committed:
Fix hot-reload-objects to build under W4 on Windows.

# Conflicts:
#	src/mono/mono/component/hot_reload.c

vargaz commented on src/mono/mono/component/hot_reload.c:
```diff
@@ -529,14 +529,14 @@ hot_reload_close_all (MonoImage *base_image)
 	for (GList *ptr = info->delta_info; ptr; ptr = ptr->next) {
-		DeltaInfo *info = (DeltaInfo *)ptr->data;
-		if (!info)
+		DeltaInfo *delta_info = (DeltaInfo *)ptr->data;
```
What was the problem here ?

lateralusX replied:
That change was mainly aligning the last DeltaInfo *info to use same naming as all others in this file, the other change in this file fix a local hiding issue,  warning C4456: declaration of 'info' hides previous local declaration, that was very common in the source and fixed in this PR.

lateralusX commented on src/mono/mono/component/debugger-agent.c:
```diff
@@ -6433,7 +6429,7 @@ get_types_for_source_file (gpointer key, gpointer value, gpointer user_data)
 			g_hash_table_insert (info->source_files, klass, files);
 
-			for (int i = 0; i < files->len; ++i) {
+			for (guint i = 0; i < files->len; ++i) {
```
Yes we probably could, did a quick test and it still builds without additional warnings. Any particular reason to do that, except that we might have usage in many places that do int i when iterating it? It will also affect the signature of g_ptr_array_remove_* that currently accept indexes to go from guint to int.

vargaz replied:
Changing it to int would reduce the amount of changes required and make it easier to write code to iterate on it.

lateralusX replied:
So for the warnings, this PR makes sure all builds as expected, but enabling more warnings might trigger more warnings related to doing unsinged->signed conversion. This PR is already very big, and I mainly did small number of changes like the one above, just to align with original types in files already touched. Could an alternative be to look into that when we enable some additional warnings that will trigger errors due to singed/unsigned mismatch? An alternative would be to revert all of the int i -> source type in this PR and then take care of all in separate PR when enabling more warnings where we also could discuss if we should change type when iterating or at the source. So mainly three options for this PR:

* Leave as is, changes done so far will simplify future fixes of warnings, but might need to be change back in a couple of places if we change any of the source types like len -> int.
* Revert the int i -> source type done in this PR (unless it breaks some other warnings enabled by this PR).
* Change len to int and fix all usage of it as part of this PR, includes changing some of the array API's as well, but as said, this PR is already big enough.

Thoughts?

lateralusX committed:
Fix build error in mscordbi socket.

lateralusX committed:
Fix additional warnings added since rebase.

lateralusX committed:
Fix Windows x86 warnings.

lateralusX committed:
Fix additional warnings.

lateralusX committed:
Fix build error.

lateralusX committed:
Fix warning build errors.

lateralusX committed:
Fix monosgen-objects and monosgen-shared to build under W4 on Windows.

# Conflicts:
#	src/mono/mono/mini/method-to-ir.c

# Conflicts:
#	src/mono/mono/mini/aot-runtime.c

lateralusX committed:
Switch back from treating all warnings as errors to promote strategy.

lateralusX committed:
Fix some more warnings.

lateralusX committed:
Disable some warnings as errors to flush out last CI failures.

lateralusX committed:
Fix debugger-objects to build under W4 on Windows.

# Conflicts:
#	src/mono/mono/component/debugger-agent.c

# Conflicts:
#	src/mono/mono/component/debugger-agent.c

lateralusX committed:
Fix warnings on x86/arm.

lateralusX committed:
Fix build error.

lateralusX committed:
Re-enable warnings.

lateralusX committed:
Fix build errors.

lateralusX commented:
All builds now pass using W4, but looks like some change (due to build of cross compiler using msvc) introduce issue on Linux Arm64 that needs further investigation.

lateralusX committed:
32-bit shift assumes wrap around in some scenarios.

akoeplinger commented on src/native/external/zlib/adler32.c:
```diff
@@ -5,6 +5,11 @@
+#ifdef _MSC_VER
+__pragma(warning (disable:4005)) /* macro redefinition */
+__pragma(warning (disable:4131)) /* uses old-style declarator */
+#endif
```
we should probably try to upstream these to https://github.com/madler/zlib so we can link the PR in https://github.com/dotnet/runtime/blob/main/src/native/external/zlib-version.txt like we do for https://github.com/dotnet/runtime/blob/main/src/native/external/libunwind-version.txt

lateralusX replied:
Something we could do after we merged this PR? Don't want to delay this PR more if not really necessary.

lateralusX replied:
Moved disabling hof zlib warnings from source files into cmake files.

lambdageek commented on src/mono/mono/metadata/class-setup-vtable.c:
```diff
@@ -1870,10 +1869,10 @@ mono_class_setup_vtable_general (MonoClass *klass, MonoMethod **overrides, int o
 				// If the slot is still empty, look in all the inherited virtual methods...
 				if ((vtable [im_slot] == NULL) && klass->parent != NULL) {
-					MonoClass *parent = klass->parent;
+					parent = klass->parent;
```
This whole assignment is redundant, isn't it?  `parent` is already set to `klass->parent` when it's declared, isn't it?  Is `klass` ever reassigned?

lateralusX replied:
Doesn't look like it, so assignment could probably be eliminated, given that klass->parent remains the same during the loop.

lambdageek commented on src/mono/mono/mini/aot-runtime.c:
```diff
@@ -3704,10 +3706,10 @@ decode_patch (MonoAotModule *aot_module, MonoMemPool *mp, MonoJumpInfo *ji, guin
 				ji->data.method = mono_get_method_checked (ref.image, ref.token, NULL, NULL, error);
 				if (!ji->data.method)
 					g_error ("AOT Runtime could not load method due to %s", mono_error_get_message (error)); /* FIXME don't swallow the error */
+				mono_error_cleanup (error);
```
This doesn't seem right. You might need to re-init `error`. (or is this really just to appease the compiler?  if `error` only has content if`!ji->data.method`, in which case we call `g_error` and abort?)

Why not use a `ERROR_DECL (local_error)` instead?

lateralusX replied:
You have the same pattern in several parts of this method, and you can reuse the error as long as it doesn't carry an error without re-init it and then it won't be reused since it will break out of switch statement and return out from function, but I agree that we should end up hitting g_error (that could abort depending on used logger) or the g_assert, so doing the cleanup might not really be needed. Normally we assert on errors or cleanup, but since this code path should get an error if it returns null and that we apparently see it as a critical error, maybe it is better to replace the condition and g_error with a mono_error_assert_msg_ok.

lambdageek commented on src/mono/mono/mini/dwarfwriter.c:
```diff
@@ -1338,10 +1338,10 @@ token_handler (MonoDisHelper *dh, MonoMethod *method, guint32 token)
 			cmethod = mono_get_method_checked (m_class_get_image (method->klass), token, NULL, NULL, error);
 			if (!cmethod)
 				g_error ("Could not load method due to %s", mono_error_get_message (error)); /* FIXME don't swallow the error */
+			mono_error_cleanup (error);
```
same as above; is the cleanup real or just to appease the compiler?

lateralusX replied:
same comment as above, error won't be reused after cleanup so no need to re-init. In this case g_error (depending on used logger implementation), might not abort and there is no assert on cmethod not being NULL, but maybe it is more clear to just use mono_error_assert_msg_ok in this case as well.

lambdageek reviewed (approved):
overall lgtm. a few questions about nits

lateralusX committed:
Review feedback.

lateralusX commented:
Rebased and fixed review feedback.

lateralusX committed:
Disable zlib warnings in cmake instead of source files.

lateralusX commented:
Moved disable of zlib warnings from source file into cmake file + rebase branch.
